designs/2025-rule-tester-assertion-options/README.md

```diff
@@ -0,0 +1,348 @@
+- Repo: https://github.com/eslint/eslint
+- Start Date: 2025-07-10
+- RFC PR: https://github.com/eslint/rfcs/pull/137
+- Authors: ST-DDT
+
+# Rule Tester Assertions Options
+
+## Summary
+
+Add options that control which assertions are required for each `RuleTester` test case.
+
+## Motivation
+
+In most eslint(-plugin)s' rules the error assertions are different from each other.
+Adding options that could be set/shared when configuring the rule tester would ensure a common base level of assertions is met throughout the (plugin's) project.
+The options could be defined on two levels. On `RuleTester`'s `constructor` effectively impacting all tests, or on the test `run` method itself, so only that set is affected.
+
+## Detailed Design
+
+### Variant 1 - Constructor based options
+
+```ts
+new RuleTester(testerConfig: {...}, assertionOptions: {
+  /**
+   * Require message assertion for each invalid test case.
+   * If `true`, errors must extend `string | RegExp | Array<{ message: string | RegExp } | { messageId: string }>`.
+   * If `'message'`, errors must extend `string | RegExp | Array<{ message: string | RegExp }>`.
+   * If `'messageId'`, errors must extend `Array<{ messageId: string }>`.
+   *
+   * @default false
+   */
+  requireMessage: boolean | 'message' | 'messageId';
+  /**
+   * Require full location assertions for each invalid test case.
+   * If `true`, errors must extend `Array<{ line: number, column: number, endLine: number, endColumn: number }>`.
```

Comment:
I think `endLine` and `endColumn` should be both of type `number | undefined`. Some core rules have tests to ensure that these properties are actually undefined, for example `unicode-bom`:

https://github.com/eslint/eslint/blob/v9.33.0/tests/lib/rules/unicode-bom.js#L43-L51

We could still require that `endLine` and `endColumn` are explicitly set in test cases depending on the `RuleTester` options.

Reply:
> To ensure that all location properties are set, `line`, `column`, `endColumn` and `endLine` should be made non-optional when `requireLocation` is `true`:

I disagree. All properties should be checked, but omitting undefined should be allowed.

Reply:
How would you go about checking an omitted property?

Reply:
````ts
if (hasOwnProperty(error, "endLine")) {
		actualLocation.endLine = message.endLine;
		expectedLocation.endLine = error.endLine;
} else if (requireLocation && hasOwnProperty(message, "endLine")) {
	assert.fail(
		"Test error must specify an 'endLine' property.",
	);
}
````

Reply:
Adjusted